File: apps/server/src/modules/warehouses/warehouse-ops.service.ts
import { prisma, type Prisma } from '@delivery/database';
import { ApiError } from '../../utils/ApiError.js';
import { buildMeta, type PaginationMeta } from '../../utils/ApiResponse.js';
import type { ParsedListQuery } from '../../utils/pagination.js';
import { eventBus, DOMAIN_EVENTS } from '../../events/eventBus.js';
import { emitToOrder, emitToAdmins, emitToUser } from '../../socket/index.js';
import { SOCKET_EVENTS } from '@delivery/types';
import { writeActivity } from '../../utils/activity.js';

const inventoryInclude = {
  package: {
    include: {
      order: {
        select: {
          id: true,
          orderNumber: true,
          status: true,
          dropoffAddress: { select: { city: true, line1: true, contactName: true } },
          customer: { select: { userId: true } },
        },
      },
    },
  },
  warehouse: { select: { id: true, code: true, name: true } },
} satisfies Prisma.WarehouseInventoryInclude;

interface ReceiveInput {
  trackingNumber: string;
  shelfCode?: string;
  zone?: string;
  note?: string;
}

interface SortInput {
  trackingNumber: string;
  shelfCode: string;
  zone?: string;
}

interface DispatchInput {
  trackingNumber: string;
  note?: string;
}

export class WarehouseOpsService {
  private async ensureWarehouse(warehouseId: string) {
    const wh = await prisma.warehouse.findUnique({ where: { id: warehouseId } });
    if (!wh) throw ApiError.notFound('Warehouse not found');
    return wh;
  }

  private async findPackageByScan(trackingNumber: string) {
    const code = trackingNumber.trim();
    const pkg = await prisma.package.findFirst({
      where: { OR: [{ trackingNumber: code }, { barcode: code }] },
      include: { order: { include: { customer: true, dropoffAddress: { select: { city: true } } } } },
    });
    if (!pkg) throw ApiError.notFound(`No parcel found for "${code}"`);
    return pkg;
  }

  private notifyOrder(
    orderId: string,
    customerUserId: string | undefined,
    orderNumber: string,
    status: string,
    body: string,
  ) {
    eventBus.publish(DOMAIN_EVENTS.ORDER_STATUS_CHANGED, { orderId, status });
    emitToOrder(orderId, SOCKET_EVENTS.ORDER_STATUS, { orderId, status });
    emitToAdmins(SOCKET_EVENTS.ORDER_STATUS, { orderId, status });
    if (customerUserId) {
      emitToUser(customerUserId, SOCKET_EVENTS.NOTIFICATION_NEW, {
        type: 'ORDER_STATUS',
        title: `Parcel ${orderNumber}`,
        body,
      });
    }
  }

  private shelfZoneForCity(city: string | null | undefined) {
    if (!city) return 'Z';
    const c = city.trim().toLowerCase();
    if (c.includes('addis')) return 'A';
    if (c.includes('hawassa')) return 'H';
    if (c.includes('adama') || c.includes('nazret')) return 'AD';
    if (c.includes('bahir')) return 'BD';
    if (c.includes('dire')) return 'DD';
    if (c.includes('mekelle')) return 'M';
    if (c.includes('gondar')) return 'G';
    return c.length >= 2 ? c.slice(0, 2).toUpperCase() : 'Z';
  }

  async receive(warehouseId: string, input: ReceiveInput, userId?: string) {
    await this.ensureWarehouse(warehouseId);
    const pkg = await this.findPackageByScan(input.trackingNumber);

    let shelfCode = input.shelfCode;
    let zone = input.zone;
    if (!shelfCode?.trim()) {
      const city = pkg.order.dropoffAddress?.city ?? 'General';
      zone = this.shelfZoneForCity(city);
      const inStock = await prisma.warehouseInventory.count({ where: { warehouseId, dispatchedAt: null } });
      shelfCode = `${zone}-${String((inStock % 20) + 1).padStart(2, '0')}`;
    }

    const inventory = await prisma.$transaction(async (tx) => {
      await tx.package.update({
        where: { id: pkg.id },
        data: { status: 'IN_WAREHOUSE', warehouseId },
      });

            if (!['DELIVERED', 'CANCELLED', 'RETURNED'].includes(pkg.order.status)) {
        await tx.order.update({ where: { id: pkg.orderId }, data: { status: 'AT_WAREHOUSE' } });
      }

      await tx.trackingEvent.create({
        data: {
          orderId: pkg.orderId,
          type: 'ARRIVED_AT_WAREHOUSE',
          status: 'Arrived at warehouse',
          description: input.note ?? `Received at warehouse${input.shelfCode ? ` · shelf ${input.shelfCode}` : ''}`,
        },
      });

            return tx.warehouseInventory.upsert({
        where: { packageId: pkg.id },
        create: { warehouseId, packageId: pkg.id, shelfCode, zone },
        update: { warehouseId, shelfCode, zone, receivedAt: new Date(), dispatchedAt: null },
        include: inventoryInclude,
      });
    });

    this.notifyOrder(
      pkg.orderId,
      pkg.order.customer?.userId,
      pkg.order.orderNumber,
      'AT_WAREHOUSE',
      `Your parcel ${pkg.order.orderNumber} arrived at the warehouse`,
    );

    if (userId) {
      await writeActivity({
        userId,
        action: 'warehouse.receive',
        metadata: { warehouseId, trackingNumber: input.trackingNumber, shelfCode },
      });
    }
    return inventory;
  }

  async sort(warehouseId: string, input: SortInput, userId?: string) {
    await this.ensureWarehouse(warehouseId);
    const pkg = await this.findPackageByScan(input.trackingNumber);

    const inv = await prisma.warehouseInventory.findUnique({ where: { packageId: pkg.id } });
    if (!inv || inv.dispatchedAt) throw ApiError.badRequest('Parcel is not currently in stock at a warehouse');

    const updated = await prisma.$transaction(async (tx) => {
      await tx.package.update({ where: { id: pkg.id }, data: { status: 'SORTED' } });
      await tx.trackingEvent.create({
        data: {
          orderId: pkg.orderId,
          type: 'SORTED',
          status: 'Sorted',
          description: `Assigned to shelf ${input.shelfCode}${input.zone ? ` (zone ${input.zone})` : ''}`,
        },
      });
      return tx.warehouseInventory.update({
        where: { packageId: pkg.id },
        data: { warehouseId, shelfCode: input.shelfCode, zone: input.zone },
        include: inventoryInclude,
      });
    });

    if (userId) {
      await writeActivity({
        userId,
        action: 'warehouse.sort',
        metadata: { warehouseId, trackingNumber: input.trackingNumber, shelfCode: input.shelfCode },
      });
    }
    return updated;
  }

  async dispatch(warehouseId: string, input: DispatchInput, userId?: string) {
    await this.ensureWarehouse(warehouseId);
    const pkg = await this.findPackageByScan(input.trackingNumber);

    const inv = await prisma.warehouseInventory.findUnique({ where: { packageId: pkg.id } });
    if (!inv || inv.dispatchedAt) throw ApiError.badRequest('Parcel is not currently in stock at a warehouse');

    const updated = await prisma.$transaction(async (tx) => {
      await tx.package.update({ where: { id: pkg.id }, data: { status: 'DISPATCHED' } });
      if (!['DELIVERED', 'CANCELLED', 'RETURNED'].includes(pkg.order.status)) {
        await tx.order.update({ where: { id: pkg.orderId }, data: { status: 'OUT_FOR_DELIVERY' } });
      }
      await tx.trackingEvent.create({
        data: {
          orderId: pkg.orderId,
          type: 'DISPATCHED',
          status: 'Dispatched',
          description: input.note ?? 'Dispatched from warehouse for delivery',
        },
      });
      return tx.warehouseInventory.update({
        where: { packageId: pkg.id },
        data: { dispatchedAt: new Date() },
        include: inventoryInclude,
      });
    });

    this.notifyOrder(
      pkg.orderId,
      pkg.order.customer?.userId,
      pkg.order.orderNumber,
      'OUT_FOR_DELIVERY',
      `Your parcel ${pkg.order.orderNumber} is out for delivery`,
    );

    if (userId) {
      await writeActivity({
        userId,
        action: 'warehouse.dispatch',
        metadata: { warehouseId, trackingNumber: input.trackingNumber },
      });
    }
    return updated;
  }

  async listInventory(
    warehouseId: string,
    query: ParsedListQuery,
    state: 'in-stock' | 'dispatched' | 'all' = 'in-stock',
  ): Promise<{ items: unknown[]; meta: PaginationMeta }> {
    await this.ensureWarehouse(warehouseId);

    const where: Prisma.WarehouseInventoryWhereInput = {
      warehouseId,
      ...(state === 'in-stock' ? { dispatchedAt: null } : {}),
      ...(state === 'dispatched' ? { dispatchedAt: { not: null } } : {}),
      ...(query.search
        ? { package: { OR: [
            { trackingNumber: { contains: query.search, mode: 'insensitive' } },
            { barcode: { contains: query.search, mode: 'insensitive' } },
          ] } }
        : {}),
    };

    const [items, total] = await Promise.all([
      prisma.warehouseInventory.findMany({
        where,
        include: inventoryInclude,
        skip: query.skip,
        take: query.take,
        orderBy: { receivedAt: query.sortOrder },
      }),
      prisma.warehouseInventory.count({ where }),
    ]);
    return { items, meta: buildMeta(query.page, query.limit, total) };
  }

  async stats(warehouseId: string) {
    const wh = await this.ensureWarehouse(warehouseId);
    const startOfDay = new Date();
    startOfDay.setHours(0, 0, 0, 0);

    const [inStock, receivedToday, dispatchedToday, byStatus, shelfGroups] = await Promise.all([
      prisma.warehouseInventory.count({ where: { warehouseId, dispatchedAt: null } }),
      prisma.warehouseInventory.count({ where: { warehouseId, receivedAt: { gte: startOfDay } } }),
      prisma.warehouseInventory.count({ where: { warehouseId, dispatchedAt: { gte: startOfDay } } }),
      prisma.package.groupBy({
        by: ['status'],
        _count: { _all: true },
        where: { warehouseId },
      }),
      prisma.warehouseInventory.groupBy({
        by: ['shelfCode'],
        where: { warehouseId, dispatchedAt: null, shelfCode: { not: null } },
        _count: { _all: true },
      }),
    ]);

    const distinctShelves = shelfGroups.length;
    const capacity = wh.capacity || 0;
    const capacityPercent = capacity > 0 ? Math.min(100, (inStock * 100) / capacity) : 0;
    const shelfUtilization =
      distinctShelves > 0 ? Math.min(100, (inStock * 100) / (distinctShelves * 10)) : 0;

    return {
      totals: {
        inStock,
        receivedToday,
        dispatchedToday,
        capacity,
        capacityPercent: Math.round(capacityPercent * 10) / 10,
        distinctShelves,
        shelfUtilization: Math.round(shelfUtilization * 10) / 10,
      },
      packagesByStatus: byStatus.map((r) => ({ status: r.status, count: r._count._all })),
    };
  }

  async inventoryByCity(warehouseId: string) {
    await this.ensureWarehouse(warehouseId);
    const items = await prisma.warehouseInventory.findMany({
      where: { warehouseId, dispatchedAt: null },
      include: inventoryInclude,
      orderBy: { receivedAt: 'asc' },
    });
    const grouped = new Map<string, unknown[]>();
    for (const item of items) {
      const city = item.package.order.dropoffAddress?.city ?? 'Unknown';
      const list = grouped.get(city) ?? [];
      list.push(item);
      grouped.set(city, list);
    }
    return Array.from(grouped.entries()).map(([city, parcels]) => ({
      city,
      count: parcels.length,
      parcels,
    }));
  }

  async agingReport(warehouseId: string) {
    await this.ensureWarehouse(warehouseId);
    const items = await prisma.warehouseInventory.findMany({
      where: { warehouseId, dispatchedAt: null },
      include: { package: { select: { trackingNumber: true } } },
      orderBy: { receivedAt: 'asc' },
    });
    const now = Date.now();
    let under24h = 0;
    let oneToThreeDays = 0;
    let threeToSevenDays = 0;
    let overSevenDays = 0;
    const stale: Array<{ trackingNumber: string; receivedAt: Date; hoursInStock: number; shelfCode: string | null }> =
      [];
    for (const inv of items) {
      const hours = (now - inv.receivedAt.getTime()) / 3_600_000;
      if (hours < 24) under24h++;
      else if (hours < 72) oneToThreeDays++;
      else if (hours < 168) threeToSevenDays++;
      else {
        overSevenDays++;
        stale.push({
          trackingNumber: inv.package.trackingNumber,
          receivedAt: inv.receivedAt,
          hoursInStock: Math.floor(hours),
          shelfCode: inv.shelfCode,
        });
      }
    }
    return {
      buckets: { under24h, oneToThreeDays, threeToSevenDays, overSevenDays },
      stale,
    };
  }

  async transfer(fromWarehouseId: string, input: { trackingNumber: string; destinationWarehouseId: string }) {
    await this.ensureWarehouse(fromWarehouseId);
    await this.ensureWarehouse(input.destinationWarehouseId);
    const pkg = await this.findPackageByScan(input.trackingNumber);
    const inv = await prisma.warehouseInventory.findUnique({ where: { packageId: pkg.id } });
    if (!inv || inv.dispatchedAt) throw ApiError.badRequest('Parcel is not currently in stock at a warehouse');
    if (inv.warehouseId !== fromWarehouseId) throw ApiError.badRequest('Parcel is not at the source warehouse');
    await prisma.warehouseInventory.update({
      where: { packageId: pkg.id },
      data: { dispatchedAt: new Date() },
    });
    return this.receive(input.destinationWarehouseId, {
      trackingNumber: input.trackingNumber,
      note: `Cross-warehouse transfer from ${fromWarehouseId}`,
    });
  }
}

export const warehouseOpsService = new WarehouseOpsService();

File: apps/guzo-api/src/main/java/et/guzo/service/WarehouseOpsService.java
package et.guzo.service;

import et.guzo.common.ApiException;
import et.guzo.common.PaginationMeta;
import et.guzo.domain.entity.*;
import et.guzo.domain.enums.OrderStatus;
import et.guzo.domain.enums.PackageStatus;
import et.guzo.domain.enums.TrackingEventType;
import et.guzo.repository.*;
import et.guzo.util.IdUtil;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.Instant;
import java.time.temporal.ChronoUnit;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

@Service
@RequiredArgsConstructor
public class WarehouseOpsService {

    private final WarehouseRepository warehouseRepository;
    private final WarehouseInventoryRepository inventoryRepository;
    private final PackageRepository packageRepository;
    private final OrderRepository orderRepository;
    private final AddressRepository addressRepository;
    private final TrackingService trackingService;

    @Transactional
    public Map<String, Object> receive(String warehouseId, Map<String, String> input) {
        ensureWarehouse(warehouseId);
        et.guzo.domain.entity.Package pkg = findPackage(input.get("trackingNumber"));
        Order order = orderRepository.findById(pkg.getOrderId()).orElseThrow(() -> ApiException.notFound("Order not found"));
        Instant now = Instant.now();
        String shelf = input.get("shelfCode");
        String zone = input.get("zone");
        if (shelf == null || shelf.isBlank()) {
            Address dropoff = addressRepository.findById(order.getDropoffAddressId()).orElse(null);
            String city = dropoff != null ? dropoff.getCity() : "General";
            zone = shelfZoneForCity(city);
            long slot = inventoryRepository.countByWarehouseIdAndDispatchedAtIsNull(warehouseId) % 20 + 1;
            shelf = zone + "-" + String.format("%02d", slot);
        }
        pkg.setStatus(PackageStatus.IN_WAREHOUSE);
        pkg.setWarehouseId(warehouseId);
        pkg.setUpdatedAt(now);
        packageRepository.save(pkg);
        if (!List.of(OrderStatus.DELIVERED, OrderStatus.CANCELLED, OrderStatus.RETURNED).contains(order.getStatus())) {
            order.setStatus(OrderStatus.AT_WAREHOUSE);
            order.setUpdatedAt(now);
            orderRepository.save(order);
        }
        trackingService.record(order.getId(), TrackingEventType.ARRIVED_AT_WAREHOUSE, OrderStatus.AT_WAREHOUSE,
            input.getOrDefault("note", "Received at warehouse"), null);
        WarehouseInventory inv = inventoryRepository.findByPackageId(pkg.getId()).orElseGet(() -> {
            WarehouseInventory row = new WarehouseInventory();
            row.setId(IdUtil.cuid());
            row.setPackageId(pkg.getId());
            row.setCreatedAt(now);
            return row;
        });
        inv.setWarehouseId(warehouseId);
        inv.setShelfCode(shelf);
        inv.setZone(zone);
        inv.setReceivedAt(now);
        inv.setDispatchedAt(null);
        inv.setUpdatedAt(now);
        return toInventoryDto(inventoryRepository.save(inv), pkg, order);
    }

    @Transactional
    public Map<String, Object> sort(String warehouseId, Map<String, String> input) {
        ensureWarehouse(warehouseId);
        et.guzo.domain.entity.Package pkg = findPackage(input.get("trackingNumber"));
        WarehouseInventory inv = inventoryRepository.findByPackageId(pkg.getId())
            .orElseThrow(() -> ApiException.badRequest("Parcel is not in warehouse inventory"));
        if (inv.getDispatchedAt() != null) throw ApiException.badRequest("Parcel already dispatched");
        Instant now = Instant.now();
        pkg.setStatus(PackageStatus.SORTED);
        pkg.setUpdatedAt(now);
        packageRepository.save(pkg);
        inv.setShelfCode(input.get("shelfCode"));
        inv.setZone(input.get("zone"));
        inv.setUpdatedAt(now);
        inventoryRepository.save(inv);
        trackingService.record(pkg.getOrderId(), TrackingEventType.SORTED, OrderStatus.AT_WAREHOUSE,
            "Assigned to shelf " + input.get("shelfCode"), null);
        Order order = orderRepository.findById(pkg.getOrderId()).orElseThrow();
        return toInventoryDto(inv, pkg, order);
    }

    @Transactional
    public Map<String, Object> dispatch(String warehouseId, Map<String, String> input) {
        ensureWarehouse(warehouseId);
        et.guzo.domain.entity.Package pkg = findPackage(input.get("trackingNumber"));
        WarehouseInventory inv = inventoryRepository.findByPackageId(pkg.getId())
            .orElseThrow(() -> ApiException.badRequest("Parcel is not in warehouse inventory"));
        if (inv.getDispatchedAt() != null) throw ApiException.badRequest("Parcel already dispatched");
        Instant now = Instant.now();
        pkg.setStatus(PackageStatus.DISPATCHED);
        pkg.setUpdatedAt(now);
        packageRepository.save(pkg);
        Order order = orderRepository.findById(pkg.getOrderId()).orElseThrow();
        if (!List.of(OrderStatus.DELIVERED, OrderStatus.CANCELLED, OrderStatus.RETURNED).contains(order.getStatus())) {
            order.setStatus(OrderStatus.OUT_FOR_DELIVERY);
            order.setUpdatedAt(now);
            orderRepository.save(order);
        }
        inv.setDispatchedAt(now);
        inv.setUpdatedAt(now);
        inventoryRepository.save(inv);
        trackingService.record(order.getId(), TrackingEventType.DISPATCHED, OrderStatus.OUT_FOR_DELIVERY,
            input.getOrDefault("note", "Dispatched from warehouse"), null);
        return toInventoryDto(inv, pkg, order);
    }

    @Transactional(readOnly = true)
    public Map<String, Object> listInventory(String warehouseId, int page, int limit, String state) {
        ensureWarehouse(warehouseId);
        PageRequest pageable = PageRequest.of(Math.max(page - 1, 0), Math.max(limit, 1), Sort.by(Sort.Direction.DESC, "receivedAt"));
        Page<WarehouseInventory> result = switch (state != null ? state : "in-stock") {
            case "dispatched" -> inventoryRepository.findByWarehouseIdAndDispatchedAtIsNotNull(warehouseId, pageable);
            case "all" -> inventoryRepository.findByWarehouseId(warehouseId, pageable);
            default -> inventoryRepository.findByWarehouseIdAndDispatchedAtIsNull(warehouseId, pageable);
        };
        List<Map<String, Object>> items = result.getContent().stream().map(inv -> {
            et.guzo.domain.entity.Package pkg = packageRepository.findById(inv.getPackageId()).orElse(null);
            Order order = pkg != null ? orderRepository.findById(pkg.getOrderId()).orElse(null) : null;
            return toInventoryDto(inv, pkg, order);
        }).toList();
        return Map.of("items", items, "meta", PaginationMeta.of(page, limit, result.getTotalElements()));
    }

    @Transactional(readOnly = true)
    public Map<String, Object> stats(String warehouseId) {
        Warehouse warehouse = warehouseRepository.findById(warehouseId)
            .orElseThrow(() -> ApiException.notFound("Warehouse not found"));
        Instant startOfDay = Instant.now().truncatedTo(ChronoUnit.DAYS);
        long inStock = inventoryRepository.countByWarehouseIdAndDispatchedAtIsNull(warehouseId);
        long receivedToday = inventoryRepository.countByWarehouseIdAndReceivedAtAfter(warehouseId, startOfDay);
        long dispatchedToday = inventoryRepository.countByWarehouseIdAndDispatchedAtAfter(warehouseId, startOfDay);
        long distinctShelves = inventoryRepository.countDistinctShelvesInStock(warehouseId);
        int capacity = warehouse.getCapacity();
        double capacityPercent = capacity > 0 ? Math.min(100.0, inStock * 100.0 / capacity) : 0;
        double shelfUtilization = distinctShelves > 0 ? Math.min(100.0, inStock * 100.0 / (distinctShelves * 10.0)) : 0;
        List<Object[]> byStatus = packageRepository.countGroupByStatusForWarehouse(warehouseId);
        Map<String, Object> totals = new LinkedHashMap<>();
        totals.put("inStock", inStock);
        totals.put("receivedToday", receivedToday);
        totals.put("dispatchedToday", dispatchedToday);
        totals.put("capacity", capacity);
        totals.put("capacityPercent", Math.round(capacityPercent * 10) / 10.0);
        totals.put("distinctShelves", distinctShelves);
        totals.put("shelfUtilization", Math.round(shelfUtilization * 10) / 10.0);
        return Map.of(
            "totals", totals,
            "packagesByStatus", byStatus.stream().map(r -> Map.of("status", r[0].toString(), "count", r[1])).toList()
        );
    }

    @Transactional(readOnly = true)
    public List<Map<String, Object>> inventoryByCity(String warehouseId) {
        ensureWarehouse(warehouseId);
        List<WarehouseInventory> rows = inventoryRepository.findInStock(warehouseId);
        Map<String, List<Map<String, Object>>> grouped = new LinkedHashMap<>();
        for (WarehouseInventory inv : rows) {
            et.guzo.domain.entity.Package pkg = packageRepository.findById(inv.getPackageId()).orElse(null);
            Order order = pkg != null ? orderRepository.findById(pkg.getOrderId()).orElse(null) : null;
            Address dropoff = order != null ? addressRepository.findById(order.getDropoffAddressId()).orElse(null) : null;
            String city = dropoff != null && dropoff.getCity() != null ? dropoff.getCity() : "Unknown";
            grouped.computeIfAbsent(city, k -> new ArrayList<>()).add(toInventoryDto(inv, pkg, order));
        }
        return grouped.entrySet().stream()
            .map(e -> Map.<String, Object>of("city", e.getKey(), "count", e.getValue().size(), "parcels", e.getValue()))
            .toList();
    }

    @Transactional(readOnly = true)
    public Map<String, Object> agingReport(String warehouseId) {
        ensureWarehouse(warehouseId);
        Instant now = Instant.now();
        List<WarehouseInventory> rows = inventoryRepository.findInStock(warehouseId);
        long under24h = 0, oneToThree = 0, threeToSeven = 0, overSeven = 0;
        List<Map<String, Object>> stale = new ArrayList<>();
        for (WarehouseInventory inv : rows) {
            long hours = ChronoUnit.HOURS.between(inv.getReceivedAt(), now);
            if (hours < 24) under24h++;
            else if (hours < 72) oneToThree++;
            else if (hours < 168) threeToSeven++;
            else {
                overSeven++;
                et.guzo.domain.entity.Package pkg = packageRepository.findById(inv.getPackageId()).orElse(null);
                if (pkg != null) {
                    Map<String, Object> row = new LinkedHashMap<>();
                    row.put("trackingNumber", pkg.getTrackingNumber());
                    row.put("receivedAt", inv.getReceivedAt());
                    row.put("hoursInStock", hours);
                    row.put("shelfCode", inv.getShelfCode());
                    stale.add(row);
                }
            }
        }
        Map<String, Object> buckets = new LinkedHashMap<>();
        buckets.put("under24h", under24h);
        buckets.put("oneToThreeDays", oneToThree);
        buckets.put("threeToSevenDays", threeToSeven);
        buckets.put("overSevenDays", overSeven);
        return Map.of("buckets", buckets, "stale", stale);
    }

    @Transactional
    public Map<String, Object> transfer(String fromWarehouseId, Map<String, String> input) {
        String toWarehouseId = input.get("destinationWarehouseId");
        if (toWarehouseId == null || toWarehouseId.isBlank()) {
            throw ApiException.badRequest("destinationWarehouseId is required");
        }
        ensureWarehouse(fromWarehouseId);
        ensureWarehouse(toWarehouseId);
        et.guzo.domain.entity.Package pkg = findPackage(input.get("trackingNumber"));
        WarehouseInventory inv = inventoryRepository.findByPackageId(pkg.getId())
            .orElseThrow(() -> ApiException.badRequest("Parcel is not in warehouse inventory"));
        if (!fromWarehouseId.equals(inv.getWarehouseId())) {
            throw ApiException.badRequest("Parcel is not at the source warehouse");
        }
        if (inv.getDispatchedAt() != null) throw ApiException.badRequest("Parcel already dispatched");
        Instant now = Instant.now();
        inv.setDispatchedAt(now);
        inv.setUpdatedAt(now);
        inventoryRepository.save(inv);
        Map<String, String> receiveInput = new LinkedHashMap<>();
        receiveInput.put("trackingNumber", pkg.getTrackingNumber());
        receiveInput.put("note", "Cross-warehouse transfer from " + fromWarehouseId);
        return receive(toWarehouseId, receiveInput);
    }

    private String shelfZoneForCity(String city) {
        if (city == null) return "Z";
        String c = city.trim().toLowerCase();
        if (c.contains("addis")) return "A";
        if (c.contains("hawassa")) return "H";
        if (c.contains("adama") || c.contains("nazret")) return "AD";
        if (c.contains("bahir")) return "BD";
        if (c.contains("dire")) return "DD";
        if (c.contains("mekelle")) return "M";
        if (c.contains("gondar")) return "G";
        return c.length() >= 2 ? c.substring(0, 2).toUpperCase() : "Z";
    }

    private void ensureWarehouse(String id) {
        warehouseRepository.findById(id).orElseThrow(() -> ApiException.notFound("Warehouse not found"));
    }

    private et.guzo.domain.entity.Package findPackage(String trackingNumber) {
        return packageRepository.findByTrackingNumber(trackingNumber.trim())
            .orElseThrow(() -> ApiException.notFound("No parcel found for that tracking number"));
    }

    private Map<String, Object> toInventoryDto(WarehouseInventory inv, et.guzo.domain.entity.Package pkg, Order order) {
        Warehouse wh = warehouseRepository.findById(inv.getWarehouseId()).orElse(null);
        Address dropoff = order != null ? addressRepository.findById(order.getDropoffAddressId()).orElse(null) : null;
        Map<String, Object> map = new LinkedHashMap<>();
        map.put("id", inv.getId());
        map.put("shelfCode", inv.getShelfCode());
        map.put("zone", inv.getZone());
        map.put("receivedAt", inv.getReceivedAt());
        map.put("dispatchedAt", inv.getDispatchedAt());
        if (wh != null) map.put("warehouse", Map.of("id", wh.getId(), "code", wh.getCode(), "name", wh.getName()));
        else map.put("warehouse", null);
        if (pkg != null) {
            Map<String, Object> pkgMap = new LinkedHashMap<>();
            pkgMap.put("id", pkg.getId());
            pkgMap.put("trackingNumber", pkg.getTrackingNumber());
            pkgMap.put("status", pkg.getStatus().name());
            pkgMap.put("description", pkg.getDescription());
            pkgMap.put("weightKg", pkg.getWeightKg());
            if (order != null) {
                Map<String, Object> orderMap = new LinkedHashMap<>();
                orderMap.put("id", order.getId());
                orderMap.put("orderNumber", order.getOrderNumber());
                orderMap.put("status", order.getStatus().name());
                if (dropoff != null) {
                    orderMap.put("dropoffAddress", Map.of(
                        "city", dropoff.getCity(), "line1", dropoff.getLine1(), "contactName", dropoff.getContactName()
                    ));
                } else orderMap.put("dropoffAddress", null);
                pkgMap.put("order", orderMap);
            } else pkgMap.put("order", null);
            map.put("package", pkgMap);
        } else map.put("package", null);
        return map;
    }
}
